src/lib/communication/DY.Lib.Communication.API.Lemmas.fst

```diff
@@ -0,0 +1,312 @@
+  (ensures (
+    let (_, tr_out) = send_confidential keys_sess_ids sender receiver payload tr in
+    trace_invariant tr_out
+  ))
+let send_confidential_proof #invs tr higher_layer_preds keys_sess_ids sender receiver payload =
+  match send_confidential keys_sess_ids sender receiver payload tr with
+  | (None, tr_out) -> ()
+  | (Some _, tr_out) -> (
+    let (Some pk_receiver, tr) = get_public_key sender keys_sess_ids.pki (PkEnc comm_layer_pkenc_tag) receiver tr in
+    let (nonce, tr) = mk_rand PkNonce (principal_label sender) 32 tr in
+    let ((), tr) = trigger_event sender (CommConfSendMsg sender receiver payload) tr in
+    let msg_encrypted = encrypt_message sender receiver payload pk_receiver nonce in
+    assert(higher_layer_preds.send_conf sender receiver payload tr);
+    assert(event_triggered tr sender (CommConfSendMsg sender receiver payload));
+    encrypt_message_proof tr sender receiver payload pk_receiver nonce;
+    let (msg_id, tr) = send_msg msg_encrypted tr in
+    assert(tr_out == tr);
+    ()
+  )
+
+val decrypt_message_proof:
+  {|cinvs:crypto_invariants|} ->
+  tr:trace ->
+  receiver:principal -> sk_receiver:bytes -> msg_encrypted:bytes ->  
+  Lemma
+  (requires
+    has_communication_layer_invariants cinvs /\
+    is_decryption_key (PkKey comm_layer_pkenc_tag empty) (principal_label receiver) tr sk_receiver /\
+    bytes_invariant tr msg_encrypted
+  )
+  (ensures (
+    match decrypt_message sk_receiver msg_encrypted with
+    | Some {sender; receiver=receiver'; payload} -> (
+      let Some msg_plain = pk_dec sk_receiver msg_encrypted in
+      (is_knowable_by (join (principal_label sender) (principal_label receiver)) tr payload /\
+        event_triggered tr sender (CommConfSendMsg sender receiver payload)) \/
+        is_publishable tr payload
+    )
+    | _ -> True
+  ))
+let decrypt_message_proof #cinvs tr receiver sk_receiver msg_encrypted =
+  match decrypt_message sk_receiver msg_encrypted with
+  | Some {sender; receiver=receiver'; payload} -> ( 
+    let Some msg_plain = pk_dec sk_receiver msg_encrypted in
+    FStar.Classical.move_requires (parse_wf_lemma communication_message (is_publishable tr)) msg_plain;
+    FStar.Classical.move_requires (parse_wf_lemma communication_message (bytes_invariant tr)) msg_plain;
+    ()
+  )
+  | _ -> ()
+
+
+val receive_confidential_proof:
+  {|invs:protocol_invariants|} ->
+  tr:trace ->
+  higher_layer_preds:comm_higher_layer_event_preds ->
+  comm_keys_ids:communication_keys_sess_ids ->
+  receiver:principal -> msg_id:timestamp ->
+  Lemma
+  (requires
+    trace_invariant tr /\ 
+    has_private_keys_invariant invs /\
+    has_communication_layer_invariants invs.crypto_invs /\
+    (
+      match receive_confidential comm_keys_ids receiver msg_id tr with
+      | (None, tr_out) -> True
+      | (Some {sender; receiver=receiver'; payload}, tr_out) ->
+        comm_higher_layer_event_preds_later tr (higher_layer_preds.receive_conf sender receiver payload)
+    ) /\
```

Comment:
I don't understand how we are supposed to use this lemma with this precondition?

Reply:
I removed the higher layer preds on the receive function. So this precondition is not needed anymore.